File: LeetCode/134.Gas_Station/Mario/134.Gas_Station.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int currGain = 0, totalGain = 0, res = 0;

        for (int i = 0; i < gas.length; i++) {
            int gain = gas[i] - cost[i];
            totalGain += gain;
            currGain += gain;

            if (currGain < 0) {
                res = i + 1;
                currGain = 0;
            }
        }

        return totalGain >= 0 ? res : -1;
    }
}

File: LeetCode/134.Gas_Station/Mario/134.Gas_Station.py
#!/usr/bin/env python3


class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        # get diff gas[i] - cost[i]
        n = len(gas)
        diff = []
        for g, c in zip(gas, cost):
            diff.append(g - c)

        # print(diff)

        cur_tank = 0
        total_tank = 0
        start = 0

        for i, d in enumerate(diff):
            cur_tank += d
            total_tank += d

            if cur_tank < 0:
                start = i + 1
                cur_tank = 0

        return start if total_tank >= 0 else -1
